File: app/src/main/java/com/niccher/mpesa_analyzer/frags/Frag_History.java
package com.niccher.mpesa_analyzer.frags;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.niccher.mpesa_analyzer.BuildConfig;
import com.niccher.mpesa_analyzer.R;
import com.niccher.mpesa_analyzer.adapter.Info_adapter;
import com.niccher.mpesa_analyzer.helpers.Prefs;
import com.niccher.mpesa_analyzer.helpers.ServiceGenerator;
import com.niccher.mpesa_analyzer.helpers.SummaryResponse;
import com.niccher.mpesa_analyzer.interfaces.JsonProcesses;
import com.niccher.mpesa_analyzer.konstants.Konstants;
import com.niccher.mpesa_analyzer.models.Mod_Summaries;

import java.security.cert.CertificateException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class Frag_History extends Fragment {

    RecyclerView recyclerView;
    TextView conn_state;
    ProgressBar conn_wait;

    JsonProcesses jsonProcesses;
    Konstants kon;
    Gson gson = null;
    ArrayList<Mod_Summaries> summariesList;
    Info_adapter summariesAdapter;

    Prefs pref;

    AppCompatActivity activity;

    public Frag_History() {}

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        kon = new Konstants();
        pref = new Prefs();

        gson = new GsonBuilder()
                .setLenient()
                .create();

        activity = (AppCompatActivity) getActivity();
        ActionBar supportActionBar = activity.getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.setTitle("Upload History");
            supportActionBar.setDisplayHomeAsUpEnabled(false);
        }
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.frag_history, container, false);

        recyclerView = view.findViewById(R.id.recy_history);
        conn_state = view.findViewById(R.id.conn_no_internet);
        conn_wait  = view.findViewById(R.id.conn_wait_internet);
        conn_wait.setVisibility(View.GONE);

        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));

        getConnectionState();
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        getConnectionState();
    }

    public void getConnectionState(){
        if (isConnected()){
            conn_wait.setVisibility(View.VISIBLE);
            conn_state.setVisibility(View.GONE);
            getSummaries();
        }else {
            conn_wait.setVisibility(View.GONE);
            recyclerView.setVisibility(View.GONE);
            isOffline("No internet connection at the moment");
        }
    }

    public boolean isConnected(){
        ConnectivityManager connectivityManager = (ConnectivityManager) getActivity().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = connectivityManager.getActiveNetworkInfo();
        return (netInfo != null && netInfo.isConnected());
    }

    private void isOffline(String msg){
        //Toast.makeText(activity, "No Internet", Toast.LENGTH_LONG).show();
        conn_state.setText(msg);
    }

    private void getSummaries() {
        conn_wait.setVisibility(View.VISIBLE);

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(kon.upload_summaries)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(ServiceGenerator.getUnsafeOkHttpClient())
                .build();

        jsonProcesses = retrofit.create(JsonProcesses.class);

        Map<String, String> parameters = new HashMap<>();
        parameters.put("varUser", pref.get_prefs_auth("auth", getContext()));
        parameters.put("varDev", pref.get_prefs_auth("print", getActivity()));

        Call <SummaryResponse> call = jsonProcesses.getSummary(parameters);
        call.enqueue(new Callback<SummaryResponse>() {
            @Override
            public void onResponse(Call<SummaryResponse> call, Response<SummaryResponse> response) {
                conn_wait.setVisibility(View.GONE);
                if(response.isSuccessful() && response.body()!=null){
                    SummaryResponse summaryResponse = response.body();
                    summariesList = new ArrayList<>(Arrays.asList(summaryResponse.getSummarizer()));
                    summariesAdapter = new Info_adapter(summariesList, getActivity());
                    recyclerView.setAdapter(summariesAdapter);
                    summariesAdapter.notifyDataSetChanged();
                }
            }

            @Override
            public void onFailure(Call<SummaryResponse> call, Throwable t) {
                conn_wait.setVisibility(View.GONE);
                isOffline("Unknown error has occurred please try again later");
                //Toast.makeText(getActivity(),  t.getMessage()+"\nUnknown error occurred, please try again", Toast.LENGTH_LONG).show();
                Log.e(kon.TAGGED, "**********************: onFailure Unknown error occurred, please try again");
                Log.e(kon.TAGGED, t.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/niccher/my_mpesa_analyzer/fragments/History/Frag_History.kt
package com.niccher.my_mpesa_analyzer.fragments.History

import android.content.Context
import android.net.ConnectivityManager
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.niccher.mpesa_analyzer.helpers.ServiceGenerators
import com.niccher.my_mpesa_analyzer.R
import com.niccher.my_mpesa_analyzer.adapter.Adapter_Frag_History
import com.niccher.my_mpesa_analyzer.databinding.FragHistoryBinding
import com.niccher.my_mpesa_analyzer.helpers.Prefs
import com.niccher.my_mpesa_analyzer.helpers.SummaryResponse
import com.niccher.my_mpesa_analyzer.interfaces.JsonProcesses
import com.niccher.my_mpesa_analyzer.konstants.Konstants
import com.niccher.my_mpesa_analyzer.models.Mod_Summaries
import kotlinx.coroutines.launch
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.ArrayList

class Frag_History : Fragment() {

    private var _binding: FragHistoryBinding? = null

    private lateinit var recyclerView: RecyclerView
    private lateinit var connState: TextView
    private lateinit var connWait: ProgressBar

    private lateinit var jsonProcesses: JsonProcesses
    private lateinit var kon: Konstants
    private lateinit var summariesAdapter: Adapter_Frag_History
    private lateinit var pref: Prefs

    private var gson: Gson = GsonBuilder().setLenient().create()
//    private var summariesList: ArrayList<Mod_Summaries> = ArrayList()
    private var summariesList: MutableList<Mod_Summaries> = mutableListOf()


    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        kon = Konstants
        pref = Prefs()

        gson = GsonBuilder()
            .setLenient()
            .create()
        
        val fragHistory = inflater.inflate(R.layout.frag_history, container, false);

        recyclerView = fragHistory.findViewById(R.id.recy_history)
        connState = fragHistory.findViewById(R.id.conn_no_internet)
        connWait = fragHistory.findViewById(R.id.conn_wait_internet)
        connWait.visibility = View.GONE
        connState.visibility = View.GONE

        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = LinearLayoutManager(requireContext())

        // Load data ONCE — uses cache when offline
        viewLifecycleOwner.lifecycleScope.launch {
            getSummaries()
        }

        return fragHistory
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setTitle("History")
    }

    override fun onResume() {
        super.onResume()
        setTitle("History")
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun setTitle(title: String) {
        // Get the activity and set the title
        val activity = requireActivity() as AppCompatActivity
        activity.supportActionBar?.title = title
        activity.supportActionBar?.setDisplayHomeAsUpEnabled(false)
    }

    private fun isOffline(msg: String) {
        connWait.visibility = View.GONE
        connState.visibility = View.VISIBLE
        connState.text = msg
        recyclerView.visibility = View.GONE
    }

    private suspend fun getSummaries() {
        try {
            connWait.visibility = View.VISIBLE

            val jsonProcesses = ServiceGenerators.createService(JsonProcesses::class.java, requireContext())
            val parameters = mapOf(
                "varUser" to pref.getPrefsAuth("auth", requireContext()),
                "varDev" to pref.getPrefsAuth("print", requireActivity())
            )

            val response = jsonProcesses.getSummary(parameters)
            summariesList = response.summarizer?.toMutableList() ?: mutableListOf()

            if (summariesList.isNotEmpty()) {
                summariesAdapter = Adapter_Frag_History(summariesList as ArrayList<Mod_Summaries>, requireActivity())
                recyclerView.adapter = summariesAdapter
                summariesAdapter.notifyDataSetChanged()

                connWait.visibility = View.GONE
                connState.visibility = View.GONE
                recyclerView.visibility = View.VISIBLE
            } else {
                isOffline("No data available")
            }

        } catch (e: Exception) {
            connWait.visibility = View.GONE
            isOffline("Offline • Showing cached data")
            Log.e(kon.TAGGED, "History: ${e.message}", e)
        }
    }

}